### PR TITLE
Wrapping RollupTable and TreeTable

### LINKED ISSUE #3249: Python wrappers for rollups and trees
With the merge of #3243, we have complete (although not entirely well tested) implementations for rollups and trees accessible from the Java API. Add wrappers from Python.

There are two areas to cover;
1. Adding the appropriate Table operations
2. Wrapping the result types (RollupTable and TreeTable, but children of HierarchicalTable)

We should focus on the former at high priority. The latter is less urgent, and will require Python API design work,

### PR DESCRIPTION
**Pull request by jmao-denver**
Fixes #3249

### CODE REVIEW

File: py/server/deephaven/rollup_table.py [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+        Raises:
+            DHError
+        """
+        try:
+            return RollupNodeOperationsRecorder(j_node_ops_recorder=self.j_rollup_table.makeNodeOperationsRecorder(
+                node_type.value))
+        except Exception as e:
+            raise DHError(e, "failed to create a RollupNodeOperationsRecorder.") from e
+
+    def with_node_operations(self, node_ops_recorders: List[RollupNodeOperationsRecorder]) -> RollupTable:
```
Comment (chipkent) [inline py/server/deephaven/rollup_table.py:150]: Is there a reasonable but shorter alternative to `node_ops_recorders` (e.g. `recorders`)?

File: py/server/deephaven/rollup_table.py [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+        """
+        try:
+            return RollupTable(
+                j_rollup_table=self.j_rollup_table.withNodeOperations(
+                    [op.j_node_ops_recorder for op in node_ops_recorders]),
+                include_constituents=self.include_constituents, aggs=self.aggs, by=self.by)
+        except Exception as e:
+            raise DHError(e, "with_node_operations on RollupTable failed.") from e
+
+    def with_filters(self, filters: Union[str, Filter, Sequence[str], Sequence[Filter]]) -> RollupTable:
```
Comment (chipkent) [inline py/server/deephaven/rollup_table.py:172]: Is there a reason this isn't named `where` to be consistent with the rest of the API?

File: py/server/deephaven/rollup_table.py [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+        try:
+            filters = to_sequence(filters)
+            if filters and isinstance(filters[0], str):
+                filters = Filter.from_(filters)
+            filters = j_array_list(filters)
+
+            return RollupTable(j_rollup_table=self.j_rollup_table.withFilters(filters),
+                               include_constituents=self.include_constituents, aggs=self.aggs, by=self.by)
+        except Exception as e:
+            raise DHError(e, "with_filters operation on RollupTable failed.") from e
```
Comment (chipkent) [inline py/server/deephaven/rollup_table.py:195]: `RollupNodeOperationsRecorder` appears to be the only outwardly facing recorder in the `RollupTable` API.  Does it make any sense to make the basic recorders private?

File: py/server/deephaven/tree_table.py [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+
+from typing import Sequence, Union
+
+import jpy
+
+from deephaven import DHError
+from deephaven._wrapper import JObjectWrapper
+from deephaven.filters import Filter
+from deephaven.jcompat import to_sequence, j_array_list
+from deephaven.rollup_table import FormatOperationsRecorder, SortOperationsRecorder, FilterOperationsRecorder
```
Comment (chipkent) [inline py/server/deephaven/tree_table.py:16]: Again, support for inclusion in `table.py`.

File: py/server/deephaven/tree_table.py [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+
+
+class TreeNodeOperationsRecorder(JObjectWrapper, FormatOperationsRecorder,
+                                 SortOperationsRecorder, FilterOperationsRecorder):
+    """Recorder for node-level operations to be applied when gathering snapshots of TreeTable. Supported operations
+    include column formatting, sorting, and filtering.
+
+    Note: It should not be instantiated directly. User code must call :meth:`~TreeTable.node_operation_recorder` to
+    create an instance of the recorder.
+    """
```
Comment (chipkent) [inline py/server/deephaven/tree_table.py:29]: pydoc is ok.  Could it be better?

File: py/server/deephaven/tree_table.py [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+    def __init__(self, j_node_ops_recorder: jpy.JType):
+        self.j_node_ops_recorder = j_node_ops_recorder
+
+
+class TreeTable(JObjectWrapper):
+    """ A TreeTable is generated as a result of applying the :meth:`~Table.tree` method on a
+    :class:`~deephaven.table.Table`.
+
+    Note: TreeTable should not be instantiated directly by user code.
+    """
```
Comment (chipkent) [inline py/server/deephaven/tree_table.py:46]: Pydoc describes how the object is created but not what it is.

File: py/server/deephaven/tree_table.py [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+        self.j_tree_table = j_tree_table
+        self.id_col = id_col
+        self.parent_col = parent_col
+
+    def node_operation_recorder(self) -> TreeNodeOperationsRecorder:
+        """Creates a TreepNodeOperationsRecorder for per-node operations to apply during snapshotting of this TreeTable.
+
+        Returns:
+            a TreeNodeOperationsRecorder
+        """
```
Comment (chipkent) [inline py/server/deephaven/tree_table.py:63]: I'm not getting a lot out of the pydocs.

File: py/server/deephaven/tree_table.py [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+
+        Args:
+            node_ops (TreeNodeOperationsRecorder): the TreeNodeOperationsRecorder containing the node operations to be
+                applied
+
+        Returns:
+            a new TreeTable
+
+        Raises:
+            DHError
```
Comment (chipkent) [inline py/server/deephaven/tree_table.py:77]: See comments in rollup table.

File: py/server/deephaven/tree_table.py [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+
+    def node_operation_recorder(self) -> TreeNodeOperationsRecorder:
+        """Creates a TreepNodeOperationsRecorder for per-node operations to apply during snapshotting of this TreeTable.
+
+        Returns:
+            a TreeNodeOperationsRecorder
+        """
+        return TreeNodeOperationsRecorder(j_node_ops_recorder=self.j_tree_table.makeNodeOperationsRecorder())
+
+    def with_node_operations(self, node_ops: TreeNodeOperationsRecorder) -> TreeTable:
```
Comment (chipkent) [inline py/server/deephaven/tree_table.py:66]: Note that the argument name is inconsistent with what is in rollup.

File: py/server/deephaven/tree_table.py [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+        """
+
+        try:
+            return TreeTable(
+                j_tree_table=self.j_tree_table.withNodeOperations(node_ops.j_node_ops_recorder),
+                id_col=self.id_col, parent_col=self.parent_col)
+        except Exception as e:
+            raise DHError(e, "with_node_operations on TreeTable failed.") from e
+
+    def with_filters(self, filters: Union[str, Filter, Sequence[str], Sequence[Filter]]) -> TreeTable:
```
Comment (chipkent) [inline py/server/deephaven/tree_table.py:87]: `where`?

File: py/server/deephaven/tree_table.py [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+        """Applies a set of filters to the columns of this TreeTable to produce and return a new TreeTable
```
Comment (chipkent) [inline py/server/deephaven/tree_table.py:88]: Compare this pydoc to what we have for `where`.  I think `where` is more clear.  https://github.com/deephaven/deephaven-core/blob/main/py/server/deephaven/table.py#L619

Also note that this and the other pydocs in this PR do not match the style of `where`.

File: py/server/tests/test_rollup_tree_table.py [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+
+from deephaven import read_csv, empty_table
+from deephaven.agg import sum_, avg, count_, first, last, max_, min_, std, abs_sum, \
+    var
+from deephaven.filters import Filter
+from deephaven.rollup_table import NodeType
+from tests.testbase import BaseTestCase
+
+
+class RollupAndTreeTableTestCase(BaseTestCase):
```
Comment (chipkent) [inline py/server/tests/test_rollup_tree_table.py:14]: In my quick look here, it looks like the code coverage is very low.

File: py/server/deephaven/table.py [hunk 13]
```diff
@@ -219,42 +220,6 @@ def _td_to_columns(table_definition):
...
-    @property
-    def j_object(self) -> jpy.JType:
-        return self.j_tree_table
-
-    def __init__(self, j_tree_table: jpy.JType, id_col: str, parent_col: str):
-        self.j_tree_table = j_tree_table
-        self.id_col = id_col
-        self.parent_col = parent_col
-
-
```
[the remark this answers is not in the document]
Comment (chipkent) [reply py/server/deephaven/table.py:257]: Splitting everything into tiny modules is not pythonic.  For example, the numpy module contains 603 things.

image

**Review by chipkent — CHANGES_REQUESTED** [review]

File: py/server/deephaven/tree_table.py (hunk 11 above)
Comment (jmao-denver) [reply py/server/deephaven/tree_table.py]: Changing it to 'where' goes against the wrapping principle of keeping the name as close to the Java one as possible. @ryan your thoughts?

File: py/server/deephaven/rollup_table.py [hunk 14]
```diff
@@ -0,0 +1,195 @@
...
+    def __init__(self, j_node_ops_recorder: jpy.JType):
+        self.j_node_ops_recorder = j_node_ops_recorder
+
+
+class RollupTable(JObjectWrapper):
+    """ A RollupTable is generated as a result of applying the :meth:`~deephaven.table.Table.rollup` operation on a
+    :class:`~deephaven.table.Table`.
+
+    Note: RollupTable should not be instantiated directly by user code.
+    """
```
[the remark this answers is not in the document]
Comment (jmao-denver) [reply py/server/deephaven/rollup_table.py]: The linked Table method actually describes what it is conceptually.

File: py/server/tests/test_rollup_tree_table.py (hunk 12 above)
Comment (jmao-denver) [reply py/server/tests/test_rollup_tree_table.py]: Actually all the supported operations and public methods are covered in the tests.

File: py/server/deephaven/tree_table.py (hunk 11 above)
Comment (rcaudy) [reply py/server/deephaven/tree_table.py]: I don't think Chip is proposing a rename of the method, just making the docs as good as the method on Table.

**Pull request closed without merging**